biscuit-parser/src/parser.rs

```diff
@@ -531,26 +531,64 @@ fn expr3(i: &str) -> IResult<&str, Expr, Error> {
 }
 
 fn expr4(i: &str) -> IResult<&str, Expr, Error> {
-    let (i, initial) = expr_term(i)?;
+    let (i, initial) = expr5(i)?;
 
-    if let Ok((i, _)) = char::<_, ()>('.')(i) {
-        let (i, op) = binary_op_4(i)?;
+    let (i, remainder) = many0(tuple((preceded(space0, binary_op_4), expr5)))(i)?;
+
+    Ok((i, fold_exprs(initial, remainder)))
+}
 
-        let (i, _) = char('(')(i)?;
-        let (i, _) = space0(i)?;
-        // we only support a single argument for now
-        let (i, arg) = expr(i)?;
-        let (i, _) = space0(i)?;
-        let (i, _) = char(')')(i)?;
+fn expr5(i: &str) -> IResult<&str, Expr, Error> {
+    alt((unary_negate, expr6))(i)
+}
 
-        let e = Expr::Binary(builder::Op::Binary(op), Box::new(initial), Box::new(arg));
+fn expr6(i: &str) -> IResult<&str, Expr, Error> {
+    let (i, initial) = expr_term(i)?;
 
-        Ok((i, e))
+    if let Ok((i, _)) = char::<_, ()>('.')(i) {
+        let bin_result = binary_method(i);
+        let un_result = unary_method(i);
+        match (bin_result, un_result) {
+            (Ok((i, (op, arg))), _) => {
+                let e = Expr::Binary(builder::Op::Binary(op), Box::new(initial), Box::new(arg));
+                Ok((i, e))
+            }
+            (_, Ok((i, op))) => {
+                let e = Expr::Unary(builder::Op::Unary(op), Box::new(initial));
+                Ok((i, e))
+            }
+            (_, Err(e)) => Err(e),
+        }
     } else {
         Ok((i, initial))
     }
 }
 
+fn binary_method(i: &str) -> IResult<&str, (builder::Binary, Expr), Error> {
+    let (i, op) = binary_op_5(i)?;
+
+    let (i, _) = char('(')(i)?;
+    let (i, _) = space0(i)?;
+    // we only support a single argument for now
+    let (i, arg) = expr(i)?;
+    let (i, _) = space0(i)?;
+    let (i, _) = char(')')(i)?;
+
+    Ok((i, (op, arg)))
+}
+
+fn unary_method(i: &str) -> IResult<&str, builder::Unary, Error> {
+    use builder::Unary;
+    let (i, op) = value(Unary::Length, tag("length"))(i)?;
+
+    let (i, _) = char('(')(i)?;
+    let (i, _) = space0(i)?;
+    let (i, _) = space0(i)?;
```

Comment:
why call `space0` twice?

Reply:
it can only be attributed to human error